### DISCUSSION TITLE
Charge Conservation and injection

### DISCUSSION THREAD

**Discussion by erny123** [epochpic/epoch#158] [Q&A] [answered]
Dear EPOCH team,

So from what I gather it seems that if the initial Electric and magnetic fields are calculated via Gauss' Law for E and B, then the FDTD algorithm will move the fields along in time and conserve charge/current. 

In EPOCH it looks like charge density is set to zero for the entire simulation so that there is no electrostatic fields even from non-neutral plasmas. 
I've been digging but haven't found a reason (mathematical/numerical) as to why this should be. Could you all explain this and why it's necessary? 

Additionally, for single species injection, how is charge conservation implemented?

Thanks,
Ernesto

**Comment by TomGoffrey** [comment, marked as answer]
I'm not really sure I understand your question. Specifically:

1. What do you mean by: "In EPOCH it looks like charge density is set to zero for the entire simulation"?
2. If a single species is injected the net charge in the simulation changes accordingly. I'm not sure what your question means.

Here's what I know:

EPOCH solves for the (change in) electric field using Ampere's law. As you say, if you have an initial charge imbalance, you should solve for the initial (electrostatic) field according to that charge imbalance, and use that field as the initial conditions. After that charge density is only a diagnostic, it doesn't play any part in the evolution of the field.

Note that the current densities are calculated from the flux of charge through cell boundaries, so that the current densities \vec{j}, are consistent with d\rho / dt = -\nabla \cdot \vec{j}, using the scheme of Esirkepov. This means that Gauss' law is automatically satisfied. To see this note from Ampere's law d / dt \nabla \cdot \vec{E} = - \nabla \cdot \vec{j} / \epsilon_0. Combined with the fact that the current densities are calculated consistently with charge conservation, this means that Gauss' law is satisfied automatically.

**Reply by erny123** [reply]
Tom, 
Thanks for the response!
So, for example, if I set an initial non-uniform spatial charge density the E-field is solved to be zero. Letting it evolve in time without any other outside disturbances from laser/fields, will keep the E-fields at zero (obviously). 
![image](https://user-images.githubusercontent.com/34816295/109709200-16cc7e80-7b51-11eb-81c2-e53f57b4c24f.png)
![image](https://user-images.githubusercontent.com/34816295/109709230-1b913280-7b51-11eb-9d8c-8877a9db682f.png)

So I'm wondering if when injecting a new particle into the simulation, if the the E-fields are recalculated from Gauss' law again inorder to take into account the introduction of new charges?

I'll take a look at the Esirkepov scheme.

**Reply by TomGoffrey** [reply]
Gauss' law is never explicitly solved in EPOCH. The only way the particles couple to the fields is through their associated current. As I said above, those currents are calculated in a manner that is consistent with charge conservation, and thus Gauss' law, but the assumption is that if you initialise EPOCH with zero field there is zero net charge. If you initialise EPOCH with a bunch of electrons, no ions, and net zero field, what you are effectively simulating is the evolution of that electron bunch within a neutralising background of stationary ions.

**Reply by erny123** [reply]
@TomGoffrey Awesome, thanks for the explanation.

**Comment by erny123** [comment]
Additionally, I'm wondering if the reason the particles injectors are restricted to the boundaries is because of this charge conservation? What is the reason keeping particles from being injected in the middle?

**Reply by TomGoffrey** [reply]
The injectors predate my work on the project. The best I can do is link another discussion which explains the motivation: https://github.com/Warwick-Plasma/epoch/issues/86#issuecomment-778163097

**Reply by erny123** [reply]
Thanks again!
